Integer in String-typed path parameter seen as invalid

### Version

4.4.1

### Context

Given a request to a path similar to this:
```
http://localhost:8080/api/v1/config/SE/111
```

The path in spec looks like:
```
/config/{countryCode}/{uniqueId}:
  get:
    summary: Some summary
    description: Some desc
    operationId: the-op
    parameters:
      - $ref: '#/components/parameters/countryCode'
      - $ref: '#/components/parameters/uniqueId'
```

And a component section as:
```
components:
  parameters:
    countryCode:
      in: path
      name: countryCode
      schema:
        type: string
        pattern: '[A-Z]{2}'
      required: true
      description: The ISO 3166-1 2-alpha country code.
    uniqueId:
      in: path
      name: uniqueId
      schema:
        type: string
      required: true
      description: The id
```

The mentioned URL breaks with the following exception:
```
DEBUG i.v.e.w.RoutingContext - RoutingContext failure (500)
io.vertx.openapi.validation.ValidatorException: The value of path parameter uniqueId is invalid. Reason: Instance type number is invalid. Expected string
	at io.vertx.openapi.validation.ValidatorException.createInvalidValue(ValidatorException.java:70)
	at io.vertx.openapi.validation.impl.RequestValidatorImpl.validateParameter(RequestValidatorImpl.java:132)
	at io.vertx.openapi.validation.impl.RequestValidatorImpl.lambda$null$2(RequestValidatorImpl.java:99)
	at io.vertx.core.impl.ContextBase.lambda$null$0(ContextBase.java:137)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
	at io.vertx.core.impl.ContextBase.lambda$executeBlocking$1(ContextBase.java:135)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: io.vertx.json.schema.JsonSchemaValidationException: Instance type number is invalid. Expected string
```

--- a/src/main/java/io/vertx/openapi/validation/impl/RequestValidatorImpl.java
+++ b/src/main/java/io/vertx/openapi/validation/impl/RequestValidatorImpl.java
@@ -122,7 +122,7 @@ public class RequestValidatorImpl extends BaseValidator implements RequestValida
     if (transformer == null) {
       throw createUnsupportedValueFormat(parameter);
     }
-    Object transformedValue = transformer.transform(parameter, value.getString());
+    Object transformedValue = transformer.transform(parameter, String.valueOf(value.get()));
     OutputUnit result = contract.getSchemaRepository().validator(parameter.getSchema()).validate(transformedValue);
 
     try {

--- a/src/main/java/io/vertx/openapi/validation/impl/ResponseValidatorImpl.java
+++ b/src/main/java/io/vertx/openapi/validation/impl/ResponseValidatorImpl.java
@@ -82,7 +82,7 @@ public class ResponseValidatorImpl extends BaseValidator implements ResponseVali
         return new RequestParameterImpl(null);
       }
     }
-    Object transformedValue = TRANSFORMER.transform(parameter, value.getString());
+    Object transformedValue = TRANSFORMER.transform(parameter, String.valueOf(value.get()));
 
     OutputUnit result = contract.getSchemaRepository().validator(parameter.getSchema()).validate(transformedValue);
 

--- a/src/main/java/io/vertx/openapi/validation/transformer/ParameterTransformer.java
+++ b/src/main/java/io/vertx/openapi/validation/transformer/ParameterTransformer.java
@@ -18,6 +18,7 @@ import io.vertx.core.json.JsonArray;
 import io.vertx.core.json.JsonObject;
 import io.vertx.openapi.contract.Parameter;
 
+import static io.vertx.json.schema.common.dsl.SchemaType.STRING;
 import static io.vertx.openapi.impl.Utils.EMPTY_JSON_ARRAY;
 import static io.vertx.openapi.impl.Utils.EMPTY_JSON_OBJECT;
 import static io.vertx.openapi.validation.ValidatorException.createCantDecodeValue;
@@ -56,6 +57,11 @@ public abstract class ParameterTransformer {
    * @return An {@link Object} holding the transformed value.
    */
   public Object transformPrimitive(Parameter parameter, String rawValue) {
+    boolean isString = STRING.equals(parameter.getSchemaType());
+    if (isString && rawValue.matches("\\w+")) {
+      return rawValue;
+    }
+
     try {
       return Json.decodeValue(rawValue);
     } catch (DecodeException de) {

--- a/src/test/java/io/vertx/openapi/validation/impl/RequestValidatorImplTest.java
+++ b/src/test/java/io/vertx/openapi/validation/impl/RequestValidatorImplTest.java
@@ -18,6 +18,7 @@ import io.vertx.core.buffer.Buffer;
 import io.vertx.core.http.HttpServerRequest;
 import io.vertx.core.json.JsonObject;
 import io.vertx.json.schema.JsonSchema;
+import io.vertx.json.schema.common.dsl.SchemaBuilder;
 import io.vertx.junit5.Checkpoint;
 import io.vertx.junit5.Timeout;
 import io.vertx.junit5.VertxExtension;
@@ -279,16 +280,33 @@ class RequestValidatorImplTest {
     assertThat(validated.getInteger()).isEqualTo(5);
   }
 
-  @Test
-  void testValidateParameterThrowInvalidValue() {
-    Parameter param = buildParam("p1", stringSchema().toJson(), false);
+  static Stream<Arguments> testValidateParameterThrowInvalidValue() {
+    return Stream.of(
+      Arguments.of(numberSchema(), true, "Instance type boolean is invalid. Expected number"),
+      Arguments.of(booleanSchema(), "{}", "Instance type object is invalid. Expected boolean"),
+      Arguments.of(booleanSchema(), "3", "Instance type number is invalid. Expected boolean")
+    );
+  }
+
+  @ParameterizedTest(name = "{index} Throw invalid value error for [{1}]")
+  @MethodSource
+  void testValidateParameterThrowInvalidValue(SchemaBuilder<?,?> schema, Object value, String reason) {
+    Parameter param = buildParam("p1", schema.toJson(), false);
     ValidatorException exception =
-      assertThrows(ValidatorException.class, () -> validator.validateParameter(param, new RequestParameterImpl("3")));
+        assertThrows(
+            ValidatorException.class,
+            () -> validator.validateParameter(param, new RequestParameterImpl(value)));
+
     assertThat(exception.type()).isEqualTo(INVALID_VALUE);
-    String reason = "Instance type number is invalid. Expected string";
-    String expectedMsg =
-      "The value of path parameter p1 is invalid. Reason: " + reason;
+    String expectedMsg = "The value of path parameter p1 is invalid. Reason: " + reason;
     assertThat(exception).hasMessageThat().isEqualTo(expectedMsg);
+  }
+
+  @Test
+  void testStringParameterValidatesNotAsIntegerType() {
+    Parameter param = buildParam("p1", stringSchema().toJson(), false);
+    RequestParameter validated = validator.validateParameter(param, new RequestParameterImpl("3"));
+    assertThat(validated.getString()).isEqualTo("3");
   }
 
   @ParameterizedTest(name = "{index} Throw error when required param is missing: {0}")

--- a/src/test/java/io/vertx/openapi/validation/impl/ResponseValidatorImplTest.java
+++ b/src/test/java/io/vertx/openapi/validation/impl/ResponseValidatorImplTest.java
@@ -18,6 +18,7 @@ import io.vertx.core.buffer.Buffer;
 import io.vertx.core.json.JsonArray;
 import io.vertx.core.json.JsonObject;
 import io.vertx.json.schema.JsonSchema;
+import io.vertx.json.schema.common.dsl.SchemaBuilder;
 import io.vertx.junit5.Checkpoint;
 import io.vertx.junit5.Timeout;
 import io.vertx.junit5.VertxExtension;
@@ -47,7 +48,9 @@ import java.util.stream.Stream;
 import static com.google.common.truth.Truth.assertThat;
 import static io.netty.handler.codec.http.HttpHeaderValues.APPLICATION_JSON;
 import static io.netty.handler.codec.http.HttpHeaderValues.TEXT_PLAIN;
+import static io.vertx.json.schema.common.dsl.Schemas.booleanSchema;
 import static io.vertx.json.schema.common.dsl.Schemas.intSchema;
+import static io.vertx.json.schema.common.dsl.Schemas.numberSchema;
 import static io.vertx.json.schema.common.dsl.Schemas.objectSchema;
 import static io.vertx.json.schema.common.dsl.Schemas.stringSchema;
 import static io.vertx.openapi.MockHelper.mockParameter;
@@ -170,15 +173,25 @@ class ResponseValidatorImplTest {
     assertThat(validated.getInteger()).isEqualTo(5);
   }
 
-  @Test
-  void testValidateParameterThrowInvalidValue() {
-    Parameter param = buildParam("p1", stringSchema().toJson(), false);
+  static Stream<Arguments> testValidateParameterThrowInvalidValue() {
+    return Stream.of(
+      Arguments.of(numberSchema(), true, "Instance type boolean is invalid. Expected number"),
+      Arguments.of(booleanSchema(), "{}", "Instance type object is invalid. Expected boolean"),
+      Arguments.of(booleanSchema(), "3", "Instance type number is invalid. Expected boolean")
+    );
+  }
+
+  @ParameterizedTest(name = "{index} Throw invalid value error for [{1}]")
+  @MethodSource
+  void testValidateParameterThrowInvalidValue(SchemaBuilder<?,?> schema, Object value, String reason) {
+    Parameter param = buildParam("p1", schema.toJson(), false);
     ValidatorException exception =
-      assertThrows(ValidatorException.class, () -> validator.validateParameter(param, new RequestParameterImpl("3")));
+      assertThrows(
+        ValidatorException.class,
+        () -> validator.validateParameter(param, new RequestParameterImpl(value)));
+
     assertThat(exception.type()).isEqualTo(INVALID_VALUE);
-    String reason = "Instance type number is invalid. Expected string";
-    String expectedMsg =
-      "The value of header parameter p1 is invalid. Reason: " + reason;
+    String expectedMsg = "The value of header parameter p1 is invalid. Reason: " + reason;
     assertThat(exception).hasMessageThat().isEqualTo(expectedMsg);
   }
 

--- a/src/test/java/io/vertx/openapi/validation/transformer/FormTransformerTest.java
+++ b/src/test/java/io/vertx/openapi/validation/transformer/FormTransformerTest.java
@@ -25,7 +25,6 @@ import org.junit.jupiter.params.provider.MethodSource;
 import java.util.stream.Stream;
 
 import static com.google.common.truth.Truth.assertThat;
-import static io.vertx.json.schema.common.dsl.Schemas.stringSchema;
 import static io.vertx.openapi.MockHelper.mockParameter;
 import static io.vertx.openapi.impl.Utils.EMPTY_JSON_ARRAY;
 import static io.vertx.openapi.impl.Utils.EMPTY_JSON_OBJECT;
@@ -33,33 +32,43 @@ import static io.vertx.openapi.contract.Location.COOKIE;
 import static io.vertx.openapi.contract.Style.FORM;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
-class FormTransformerTest {
-  private static final Parameter DUMMY_PARAM = mockFormParameter("dummy", false);
-  private static final Parameter DUMMY_PARAM_EXPLODE = mockFormParameter("dummy", true);
+class FormTransformerTest implements SchemaSupport {
+
+  private static final Parameter OBJECT_PARAM = mockFormParameter(OBJECT_SCHEMA, false);
+  private static final Parameter ARRAY_PARAM = mockFormParameter(ARRAY_SCHEMA, false);
+  private static final Parameter STRING_PARAM = mockFormParameter(STRING_SCHEMA, false);
+  private static final Parameter NUMBER_PARAM = mockFormParameter(NUMBER_SCHEMA, false);
+  private static final Parameter INTEGER_PARAM = mockFormParameter(INTEGER_SCHEMA, false);
+  private static final Parameter BOOLEAN_PARAM = mockFormParameter(BOOLEAN_SCHEMA, false);
+
+  private static final Parameter OBJECT_PARAM_EXPLODE = mockFormParameter(OBJECT_SCHEMA, true);
+  private static final Parameter ARRAY_PARAM_EXPLODE = mockFormParameter(ARRAY_SCHEMA, true);
+
   private static final FormTransformer TRANSFORMER = new FormTransformer();
 
-  private static Parameter mockFormParameter(String name, boolean explode) {
-    return mockParameter(name, COOKIE, FORM, explode, JsonSchema.of(stringSchema().toJson()));
+  private static Parameter mockFormParameter(JsonSchema schema, boolean explode) {
+    return mockParameter(NAME, COOKIE, FORM, explode, schema);
   }
 
   private static Stream<Arguments> provideValidPrimitiveValues() {
     return Stream.of(
-      Arguments.of("(String) empty", "", ""),
-      Arguments.of("(String) \"foobar\"", "foobar", "foobar"),
-      Arguments.of("(Number) 14.6767", "14.6767", 14.6767),
-      Arguments.of("(Integer) 42", "42", 42),
-      Arguments.of("(Boolean) true", "true", true)
+      Arguments.of("(String) empty", STRING_PARAM, "", ""),
+      Arguments.of("(String) 42", STRING_PARAM, "42", "42"),
+      Arguments.of("(String) \"foobar\"", STRING_PARAM, "foobar", "foobar"),
+      Arguments.of("(Number) 14.6767", NUMBER_PARAM, "14.6767", 14.6767),
+      Arguments.of("(Integer) 42", INTEGER_PARAM, "42", 42),
+      Arguments.of("(Boolean) true", BOOLEAN_PARAM, "true", true)
     );
   }
 
   private static Stream<Arguments> provideValidArrayValues() {
     JsonArray expectedComplex = new JsonArray().add("Hello").add(1).add(false).add(13.37);
     return Stream.of(
-      Arguments.of("empty", DUMMY_PARAM, "", EMPTY_JSON_ARRAY),
-      Arguments.of("3", DUMMY_PARAM, "3", new JsonArray().add(3)),
-      Arguments.of("dummy=3 (exploded)", DUMMY_PARAM_EXPLODE, "dummy=3", new JsonArray().add(3)),
-      Arguments.of("Hello,1,false,13.37", DUMMY_PARAM, "Hello,1,false,13.37", expectedComplex),
-      Arguments.of("dummy=Hello&dummy=1&dummy=false&dummy=13.37 (exploded)", DUMMY_PARAM_EXPLODE,
+      Arguments.of("empty", ARRAY_PARAM, "", EMPTY_JSON_ARRAY),
+      Arguments.of("3", ARRAY_PARAM, "3", new JsonArray().add(3)),
+      Arguments.of("dummy=3 (exploded)", ARRAY_PARAM_EXPLODE, "dummy=3", new JsonArray().add(3)),
+      Arguments.of("Hello,1,false,13.37", ARRAY_PARAM, "Hello,1,false,13.37", expectedComplex),
+      Arguments.of("dummy=Hello&dummy=1&dummy=false&dummy=13.37 (exploded)", ARRAY_PARAM_EXPLODE,
         "dummy=Hello&dummy=1&dummy=false&dummy=13.37", expectedComplex)
     );
   }
@@ -71,17 +80,17 @@ class FormTransformerTest {
       new JsonObject().put("string", "foo").put("integer", 42).put("boolean", true).put("number", 13.37);
 
     return Stream.of(
-      Arguments.of("empty", DUMMY_PARAM, "", EMPTY_JSON_OBJECT),
-      Arguments.of("empty (exploded)", DUMMY_PARAM_EXPLODE, "", EMPTY_JSON_OBJECT),
-      Arguments.of(complexRaw, DUMMY_PARAM, complexRaw, expected),
-      Arguments.of(complexExplodedRaw + " (exploded)", DUMMY_PARAM_EXPLODE, complexExplodedRaw, expected)
+      Arguments.of("empty", OBJECT_PARAM, "", EMPTY_JSON_OBJECT),
+      Arguments.of("empty (exploded)", OBJECT_PARAM_EXPLODE, "", EMPTY_JSON_OBJECT),
+      Arguments.of(complexRaw, OBJECT_PARAM, complexRaw, expected),
+      Arguments.of(complexExplodedRaw + " (exploded)", OBJECT_PARAM_EXPLODE, complexExplodedRaw, expected)
     );
   }
 
   @ParameterizedTest(name = "{index} Transform \"Cookie\" parameter of style \"from\" with primitive value: {0}")
   @MethodSource("provideValidPrimitiveValues")
-  void testTransformPrimitiveValid(String scenario, String rawValue, Object expectedValue) {
-    assertThat(TRANSFORMER.transformPrimitive(DUMMY_PARAM, rawValue)).isEqualTo(expectedValue);
+  void testTransformPrimitiveValid(String scenario, Parameter parameter, String rawValue, Object expectedValue) {
+    assertThat(TRANSFORMER.transformPrimitive(parameter, rawValue)).isEqualTo(expectedValue);
   }
 
   @ParameterizedTest(name = "{index} Transform \"Cookie\" parameter of style \"from\" with array value: {0}")
@@ -100,7 +109,7 @@ class FormTransformerTest {
   void testInvalidValues() {
     String invalidObject = "string,foo,number";
     ValidatorException exception =
-      assertThrows(ValidatorException.class, () -> TRANSFORMER.transformObject(DUMMY_PARAM, invalidObject));
+      assertThrows(ValidatorException.class, () -> TRANSFORMER.transformObject(STRING_PARAM, invalidObject));
     String expectedMsg = "The formatting of the value of cookie parameter dummy doesn't match to style form.";
     assertThat(exception).hasMessageThat().isEqualTo(expectedMsg);
   }

--- a/src/test/java/io/vertx/openapi/validation/transformer/LabelTransformerTest.java
+++ b/src/test/java/io/vertx/openapi/validation/transformer/LabelTransformerTest.java
@@ -27,8 +27,6 @@ import org.junit.jupiter.params.provider.ValueSource;
 import java.util.stream.Stream;
 
 import static com.google.common.truth.Truth.assertThat;
-import static io.vertx.json.schema.common.dsl.Schemas.arraySchema;
-import static io.vertx.json.schema.common.dsl.Schemas.stringSchema;
 import static io.vertx.openapi.MockHelper.mockParameter;
 import static io.vertx.openapi.impl.Utils.EMPTY_JSON_ARRAY;
 import static io.vertx.openapi.impl.Utils.EMPTY_JSON_OBJECT;
@@ -38,23 +36,32 @@ import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.verify;
 
-class LabelTransformerTest {
-  private static final Parameter DUMMY_PARAM = mockLabelParameter("dummy", false);
-  private static final Parameter DUMMY_PARAM_EXPLODE = mockLabelParameter("dummy", true);
+class LabelTransformerTest implements SchemaSupport {
+
+  private static final Parameter OBJECT_PARAM = mockLabelParameter(OBJECT_SCHEMA, false);
+  private static final Parameter ARRAY_PARAM = mockLabelParameter(ARRAY_SCHEMA, false);
+  private static final Parameter STRING_PARAM = mockLabelParameter(STRING_SCHEMA, false);
+  private static final Parameter NUMBER_PARAM = mockLabelParameter(NUMBER_SCHEMA, false);
+  private static final Parameter INTEGER_PARAM = mockLabelParameter(INTEGER_SCHEMA, false);
+  private static final Parameter BOOLEAN_PARAM = mockLabelParameter(BOOLEAN_SCHEMA, false);
+
+  private static final Parameter OBJECT_PARAM_EXPLODE = mockLabelParameter(OBJECT_SCHEMA, true);
+  private static final Parameter ARRAY_PARAM_EXPLODE = mockLabelParameter(ARRAY_SCHEMA, true);
 
   private static final LabelTransformer TRANSFORMER = new LabelTransformer();
 
-  private static Parameter mockLabelParameter(String name, boolean explode) {
-    return mockParameter(name, PATH, LABEL, explode, JsonSchema.of(stringSchema().toJson()));
+  private static Parameter mockLabelParameter(JsonSchema schema, boolean explode) {
+    return mockParameter(NAME, PATH, LABEL, explode, schema);
   }
 
   private static Stream<Arguments> provideValidPrimitiveValues() {
     return Stream.of(
-      Arguments.of("(String) empty", ".", ""),
-      Arguments.of("(String) \".foobar\"", ".foobar", "foobar"),
-      Arguments.of("(Number) .14.6767", ".14.6767", 14.6767),
-      Arguments.of("(Integer) .42", ".42", 42),
-      Arguments.of("(Boolean) .true", ".true", true)
+      Arguments.of("(String) empty", STRING_PARAM, ".", ""),
+      Arguments.of("(String) .44", STRING_PARAM, ".44", "44"),
+      Arguments.of("(String) \".foobar\"", STRING_PARAM, ".foobar", "foobar"),
+      Arguments.of("(Number) .14.6767", NUMBER_PARAM, ".14.6767", 14.6767),
+      Arguments.of("(Integer) .42", INTEGER_PARAM, ".42", 42),
+      Arguments.of("(Boolean) .true", BOOLEAN_PARAM, ".true", true)
     );
   }
 
@@ -62,11 +69,11 @@ class LabelTransformerTest {
     JsonArray expectedNoNumber = new JsonArray().add("Hello").add(1).add(false);
     JsonArray expectedComplex = expectedNoNumber.copy().add(13.37);
     return Stream.of(
-      Arguments.of("empty .", DUMMY_PARAM, ".", EMPTY_JSON_ARRAY),
-      Arguments.of(".3", DUMMY_PARAM, ".3", new JsonArray().add(3)),
-      Arguments.of(".3 (exploded)", DUMMY_PARAM_EXPLODE, ".3", new JsonArray().add(3)),
-      Arguments.of(".Hello,1,false,13.37", DUMMY_PARAM, ".Hello,1,false,13.37", expectedComplex),
-      Arguments.of(".Hello.1.false (exploded)", DUMMY_PARAM_EXPLODE, ".Hello.1.false", expectedNoNumber)
+      Arguments.of("empty .", ARRAY_PARAM, ".", EMPTY_JSON_ARRAY),
+      Arguments.of(".3", ARRAY_PARAM, ".3", new JsonArray().add(3)),
+      Arguments.of(".3 (exploded)", ARRAY_PARAM_EXPLODE, ".3", new JsonArray().add(3)),
+      Arguments.of(".Hello,1,false,13.37", ARRAY_PARAM, ".Hello,1,false,13.37", expectedComplex),
+      Arguments.of(".Hello.1.false (exploded)", ARRAY_PARAM_EXPLODE, ".Hello.1.false", expectedNoNumber)
     );
   }
 
@@ -78,19 +85,19 @@ class LabelTransformerTest {
     JsonObject expectedComplex = expectedNoNumber.copy().put("number", 13.37);
 
     return Stream.of(
-      Arguments.of("empty", DUMMY_PARAM, ".", EMPTY_JSON_OBJECT),
-      Arguments.of("empty (exploded)", DUMMY_PARAM_EXPLODE, ".", EMPTY_JSON_OBJECT),
-      Arguments.of(complexRaw, DUMMY_PARAM, complexRaw, expectedComplex),
-      Arguments.of(complexExplodedRaw + " (exploded)", DUMMY_PARAM_EXPLODE, complexExplodedRaw,
+      Arguments.of("empty", OBJECT_PARAM, ".", EMPTY_JSON_OBJECT),
+      Arguments.of("empty (exploded)", OBJECT_PARAM_EXPLODE, ".", EMPTY_JSON_OBJECT),
+      Arguments.of(complexRaw, OBJECT_PARAM, complexRaw, expectedComplex),
+      Arguments.of(complexExplodedRaw + " (exploded)", OBJECT_PARAM_EXPLODE, complexExplodedRaw,
         expectedNoNumber)
     );
   }
 
   @ParameterizedTest(name = "{index} Transform \"Path\" parameter of style \"label\" with primitive value: {0}")
   @MethodSource("provideValidPrimitiveValues")
-  void testTransformPrimitiveValid(String scenario, String rawValue, Object expectedValue) {
+  void testTransformPrimitiveValid(String scenario, Parameter parameter, String rawValue, Object expectedValue) {
     // Leading dot will be removed in transform method
-    assertThat(TRANSFORMER.transformPrimitive(DUMMY_PARAM, rawValue.substring(1))).isEqualTo(expectedValue);
+    assertThat(TRANSFORMER.transformPrimitive(parameter, rawValue.substring(1))).isEqualTo(expectedValue);
   }
 
   @ParameterizedTest(name = "{index} Transform \"Path\" parameter of style \"label\" with array value: {0}")
@@ -111,7 +118,7 @@ class LabelTransformerTest {
   void testInvalidValues() {
     String invalidObject = ".string,foo,number";
     ValidatorException exception =
-      assertThrows(ValidatorException.class, () -> TRANSFORMER.transformObject(DUMMY_PARAM, invalidObject));
+      assertThrows(ValidatorException.class, () -> TRANSFORMER.transformObject(OBJECT_PARAM, invalidObject));
     String expectedMsg = "The formatting of the value of path parameter dummy doesn't match to style label.";
     assertThat(exception).hasMessageThat().isEqualTo(expectedMsg);
   }
@@ -120,7 +127,7 @@ class LabelTransformerTest {
   @ValueSource(strings = {"", "notStartingWithDot"})
   void testTransformException(String invalidValue) {
     ValidatorException exception =
-      assertThrows(ValidatorException.class, () -> TRANSFORMER.transform(DUMMY_PARAM, invalidValue));
+      assertThrows(ValidatorException.class, () -> TRANSFORMER.transform(STRING_PARAM, invalidValue));
     String expectedMsg = "The formatting of the value of path parameter dummy doesn't match to style label.";
     assertThat(exception).hasMessageThat().isEqualTo(expectedMsg);
   }
@@ -128,10 +135,8 @@ class LabelTransformerTest {
   @Test
   @DisplayName("Ensure that values with leading dot are forwarded to the related transform methods")
   void testTransform() {
-    JsonSchema schema = JsonSchema.of(arraySchema().toJson());
-    Parameter param = mockParameter("dummy", PATH, LABEL, false, schema);
     LabelTransformer spyTransformer = spy(new LabelTransformer());
-    spyTransformer.transform(param, ".5");
-    verify(spyTransformer).transform(param, ".5");
+    spyTransformer.transform(STRING_PARAM, ".5");
+    verify(spyTransformer).transform(STRING_PARAM, ".5");
   }
 }

--- a/src/test/java/io/vertx/openapi/validation/transformer/MatrixTransformerTest.java
+++ b/src/test/java/io/vertx/openapi/validation/transformer/MatrixTransformerTest.java
@@ -39,37 +39,42 @@ import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.verify;
 
-class MatrixTransformerTest {
-  private static final Parameter DUMMY_PARAM = mockMatrixParameter("dummy", false, stringSchema().toJson());
-  private static final Parameter DUMMY_PARAM_EXPLODE = mockMatrixParameter("dummy", true, stringSchema().toJson());
+class MatrixTransformerTest implements SchemaSupport {
 
-  private static final Parameter DUMMY_PARAM_OBJECT_EXPLODE =
-    mockMatrixParameter("dummy", true, objectSchema().toJson());
+  private static final Parameter OBJECT_PARAM = mockMatrixParameter(OBJECT_SCHEMA, false);
+  private static final Parameter ARRAY_PARAM = mockMatrixParameter(ARRAY_SCHEMA, false);
+  private static final Parameter STRING_PARAM = mockMatrixParameter(STRING_SCHEMA, false);
+  private static final Parameter NUMBER_PARAM = mockMatrixParameter(NUMBER_SCHEMA, false);
+  private static final Parameter INTEGER_PARAM = mockMatrixParameter(INTEGER_SCHEMA, false);
+  private static final Parameter BOOLEAN_PARAM = mockMatrixParameter(BOOLEAN_SCHEMA, false);
+
+  private static final Parameter OBJECT_PARAM_EXPLODE = mockMatrixParameter(OBJECT_SCHEMA, true);
+  private static final Parameter ARRAY_PARAM_EXPLODE = mockMatrixParameter(ARRAY_SCHEMA, true);
 
   private static final MatrixTransformer TRANSFORMER = new MatrixTransformer();
 
-  private static Parameter mockMatrixParameter(String name, boolean explode, JsonObject schema) {
-    return mockParameter(name, PATH, MATRIX, explode, JsonSchema.of(schema));
+  private static Parameter mockMatrixParameter(JsonSchema schema, boolean explode) {
+    return mockParameter(NAME, PATH, MATRIX, explode, schema);
   }
 
   private static Stream<Arguments> provideValidPrimitiveValues() {
     return Stream.of(
-      Arguments.of("(String) empty", ";dummy=", ""),
-      Arguments.of("(String) \";dummy=foobar\"", ";dummy=foobar", "foobar"),
-      Arguments.of("(Number) ;dummy=14.6767", ";dummy=14.6767", 14.6767),
-      Arguments.of("(Integer) ;dummy=42", ";dummy=42", 42),
-      Arguments.of("(Boolean) ;dummy=true", ";dummy=true", true)
+      Arguments.of("(String) empty", STRING_PARAM, ";dummy=", ""),
+      Arguments.of("(String) \";dummy=foobar\"", STRING_PARAM, ";dummy=foobar", "foobar"),
+      Arguments.of("(Number) ;dummy=14.6767", NUMBER_PARAM, ";dummy=14.6767", 14.6767),
+      Arguments.of("(Integer) ;dummy=42", INTEGER_PARAM, ";dummy=42", 42),
+      Arguments.of("(Boolean) ;dummy=true", BOOLEAN_PARAM, ";dummy=true", true)
     );
   }
 
   private static Stream<Arguments> provideValidArrayValues() {
     JsonArray expectedComplex = new JsonArray().add("Hello").add(1).add(false).add(13.37);
     return Stream.of(
-      Arguments.of("empty ;dummy=", DUMMY_PARAM, ";dummy=", EMPTY_JSON_ARRAY),
-      Arguments.of(";dummy=3", DUMMY_PARAM, ";dummy=3", new JsonArray().add(3)),
-      Arguments.of(";dummy=3 (exploded)", DUMMY_PARAM_EXPLODE, ";dummy=3", new JsonArray().add(3)),
-      Arguments.of(";dummy=Hello,1,false,13.37", DUMMY_PARAM, ";dummy=Hello,1,false,13.37", expectedComplex),
-      Arguments.of(";dummy=Hello;dummy=1;dummy=false;dummy=13.37 (exploded)", DUMMY_PARAM_EXPLODE,
+      Arguments.of("empty ;dummy=", ARRAY_PARAM, ";dummy=", EMPTY_JSON_ARRAY),
+      Arguments.of(";dummy=3", ARRAY_PARAM, ";dummy=3", new JsonArray().add(3)),
+      Arguments.of(";dummy=3 (exploded)", ARRAY_PARAM_EXPLODE, ";dummy=3", new JsonArray().add(3)),
+      Arguments.of(";dummy=Hello,1,false,13.37", ARRAY_PARAM, ";dummy=Hello,1,false,13.37", expectedComplex),
+      Arguments.of(";dummy=Hello;dummy=1;dummy=false;dummy=13.37 (exploded)", ARRAY_PARAM_EXPLODE,
         ";dummy=Hello;dummy=1;dummy=false;dummy=13.37", expectedComplex)
     );
   }
@@ -81,19 +86,19 @@ class MatrixTransformerTest {
       new JsonObject().put("string", "foo").put("integer", 42).put("boolean", true).put("number", 13.37);
 
     return Stream.of(
-      Arguments.of("empty", DUMMY_PARAM, ";dummy=", EMPTY_JSON_OBJECT),
-      Arguments.of("empty (exploded)", DUMMY_PARAM_OBJECT_EXPLODE, ";", EMPTY_JSON_OBJECT),
-      Arguments.of(complexRaw, DUMMY_PARAM, complexRaw, expected),
-      Arguments.of(complexExplodedRaw + " (exploded)", DUMMY_PARAM_OBJECT_EXPLODE, complexExplodedRaw, expected)
+      Arguments.of("empty", OBJECT_PARAM, ";dummy=", EMPTY_JSON_OBJECT),
+      Arguments.of("empty (exploded)", OBJECT_PARAM_EXPLODE, ";", EMPTY_JSON_OBJECT),
+      Arguments.of(complexRaw, OBJECT_PARAM, complexRaw, expected),
+      Arguments.of(complexExplodedRaw + " (exploded)", OBJECT_PARAM_EXPLODE, complexExplodedRaw, expected)
     );
   }
 
   @ParameterizedTest(name = "{index} Transform \"Path\" parameter of style \"matrix\" with primitive value: {0}")
   @MethodSource("provideValidPrimitiveValues")
-  void testTransformPrimitiveValid(String scenario, String rawValue, Object expectedValue) {
+  void testTransformPrimitiveValid(String scenario, Parameter parameter, String rawValue, Object expectedValue) {
     // Leading prefix will be removed in transform method
-    int prefixLength = TRANSFORMER.buildPrefix(DUMMY_PARAM).length();
-    assertThat(TRANSFORMER.transformPrimitive(DUMMY_PARAM, rawValue.substring(prefixLength))).isEqualTo(expectedValue);
+    int prefixLength = TRANSFORMER.buildPrefix(parameter).length();
+    assertThat(TRANSFORMER.transformPrimitive(parameter, rawValue.substring(prefixLength))).isEqualTo(expectedValue);
   }
 
   @ParameterizedTest(name = "{index} Transform \"Path\" parameter of style \"matrix\" with array value: {0}")
@@ -116,7 +121,7 @@ class MatrixTransformerTest {
   void testInvalidValues() {
     String invalidObject = ";dummy=string,foo,number";
     ValidatorException exception =
-      assertThrows(ValidatorException.class, () -> TRANSFORMER.transformObject(DUMMY_PARAM, invalidObject));
+      assertThrows(ValidatorException.class, () -> TRANSFORMER.transformObject(OBJECT_PARAM, invalidObject));
     String expectedMsg = "The formatting of the value of path parameter dummy doesn't match to style matrix.";
     assertThat(exception).hasMessageThat().isEqualTo(expectedMsg);
   }
@@ -125,7 +130,7 @@ class MatrixTransformerTest {
   @ValueSource(strings = {"", ";dummy", ";foo="})
   void testTransformException(String invalidValue) {
     ValidatorException exception =
-      assertThrows(ValidatorException.class, () -> TRANSFORMER.transform(DUMMY_PARAM, invalidValue));
+      assertThrows(ValidatorException.class, () -> TRANSFORMER.transform(STRING_PARAM, invalidValue));
     String expectedMsg = "The formatting of the value of path parameter dummy doesn't match to style matrix.";
     assertThat(exception).hasMessageThat().isEqualTo(expectedMsg);
   }
@@ -133,10 +138,8 @@ class MatrixTransformerTest {
   @Test
   @DisplayName("Ensure that values with leading matrix prefix are forwarded to the related transform methods")
   void testTransform() {
-    JsonSchema schema = JsonSchema.of(arraySchema().toJson());
-    Parameter param = mockParameter("dummy", PATH, MATRIX, false, schema);
     MatrixTransformer spyTransformer = spy(new MatrixTransformer());
-    spyTransformer.transform(param, ";dummy=5");
-    verify(spyTransformer).transform(param, ";dummy=5");
+    spyTransformer.transform(ARRAY_PARAM, ";dummy=5");
+    verify(spyTransformer).transform(ARRAY_PARAM, ";dummy=5");
   }
 }

--- a/src/test/java/io/vertx/openapi/validation/transformer/SchemaSupport.java
+++ b/src/test/java/io/vertx/openapi/validation/transformer/SchemaSupport.java
@@ -1,0 +1,22 @@
+package io.vertx.openapi.validation.transformer;
+
+import static io.vertx.json.schema.common.dsl.Schemas.arraySchema;
+import static io.vertx.json.schema.common.dsl.Schemas.booleanSchema;
+import static io.vertx.json.schema.common.dsl.Schemas.intSchema;
+import static io.vertx.json.schema.common.dsl.Schemas.numberSchema;
+import static io.vertx.json.schema.common.dsl.Schemas.objectSchema;
+import static io.vertx.json.schema.common.dsl.Schemas.stringSchema;
+
+import io.vertx.json.schema.JsonSchema;
+
+interface SchemaSupport {
+
+  String NAME = "dummy";
+
+  JsonSchema OBJECT_SCHEMA = JsonSchema.of(objectSchema().toJson());
+  JsonSchema ARRAY_SCHEMA = JsonSchema.of(arraySchema().toJson());
+  JsonSchema STRING_SCHEMA = JsonSchema.of(stringSchema().toJson());
+  JsonSchema NUMBER_SCHEMA = JsonSchema.of(numberSchema().toJson());
+  JsonSchema INTEGER_SCHEMA = JsonSchema.of(intSchema().toJson());
+  JsonSchema BOOLEAN_SCHEMA = JsonSchema.of(booleanSchema().toJson());
+}

--- a/src/test/java/io/vertx/openapi/validation/transformer/SimpleTransformerTest.java
+++ b/src/test/java/io/vertx/openapi/validation/transformer/SimpleTransformerTest.java
@@ -38,36 +38,40 @@ import static io.vertx.openapi.contract.Location.PATH;
 import static io.vertx.openapi.contract.Style.SIMPLE;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
-class SimpleTransformerTest {
-  private static final Parameter DUMMY_PARAM = mockSimpleParameter("dummy", false);
-  private static final Parameter DUMMY_PARAM_EXPLODE = mockSimpleParameter("dummy", true);
+class SimpleTransformerTest implements SchemaSupport {
+
+  private static final Parameter OBJECT_PARAM = mockSimpleParameter(OBJECT_SCHEMA, false);
+  private static final Parameter ARRAY_PARAM = mockSimpleParameter(ARRAY_SCHEMA, false);
+  private static final Parameter STRING_PARAM = mockSimpleParameter(STRING_SCHEMA, false);
+  private static final Parameter NUMBER_PARAM = mockSimpleParameter(NUMBER_SCHEMA, false);
+  private static final Parameter INTEGER_PARAM = mockSimpleParameter(INTEGER_SCHEMA, false);
+  private static final Parameter BOOLEAN_PARAM = mockSimpleParameter(BOOLEAN_SCHEMA, false);
+
+  private static final Parameter OBJECT_PARAM_EXPLODE = mockSimpleParameter(OBJECT_SCHEMA, true);
 
   private static final SimpleTransformer TRANSFORMER = new SimpleTransformer();
 
-  private static Parameter mockSimpleParameter(String name, boolean explode) {
-    return mockParameter(name, PATH, SIMPLE, explode, JsonSchema.of(stringSchema().toJson()));
-  }
-
-  private static Parameter buildSimplePathParameter(SchemaBuilder<?, ?> schema) {
-    return mockParameter("dummy", PATH, SIMPLE, false, JsonSchema.of(schema.toJson()));
+  private static Parameter mockSimpleParameter(JsonSchema schema, boolean explode) {
+    return mockParameter(NAME, PATH, SIMPLE, explode, schema);
   }
 
   private static Stream<Arguments> provideValidPrimitiveValues() {
     return Stream.of(
-      Arguments.of("(String) empty", buildSimplePathParameter(stringSchema()), "", ""),
-      Arguments.of("(String) \"foobar\"", buildSimplePathParameter(stringSchema()), "foobar", "foobar"),
-      Arguments.of("(Number) 14.6767", buildSimplePathParameter(numberSchema()), "14.6767", 14.6767),
-      Arguments.of("(Integer) 42", buildSimplePathParameter(intSchema()), "42", 42),
-      Arguments.of("(Boolean) true", buildSimplePathParameter(booleanSchema()), "true", true)
+      Arguments.of("(String) empty", STRING_PARAM, "", ""),
+      Arguments.of("(String) 44", STRING_PARAM, "44", "44"),
+      Arguments.of("(String) \"foobar\"", STRING_PARAM, "foobar", "foobar"),
+      Arguments.of("(Number) 14.6767", NUMBER_PARAM, "14.6767", 14.6767),
+      Arguments.of("(Integer) 42", INTEGER_PARAM, "42", 42),
+      Arguments.of("(Boolean) true", BOOLEAN_PARAM, "true", true)
     );
   }
 
   private static Stream<Arguments> provideValidArrayValues() {
     JsonArray expectedComplex = new JsonArray().add("Hello").add(13.37).add(1).add(false);
     return Stream.of(
-      Arguments.of("empty", DUMMY_PARAM, "", EMPTY_JSON_ARRAY),
-      Arguments.of("3", DUMMY_PARAM, "3", new JsonArray().add(3)),
-      Arguments.of("Hello,13.37,1,false", DUMMY_PARAM, "Hello,13.37,1,false", expectedComplex)
+      Arguments.of("empty", ARRAY_PARAM, "", EMPTY_JSON_ARRAY),
+      Arguments.of("3", ARRAY_PARAM, "3", new JsonArray().add(3)),
+      Arguments.of("Hello,13.37,1,false", ARRAY_PARAM, "Hello,13.37,1,false", expectedComplex)
     );
   }
 
@@ -78,10 +82,10 @@ class SimpleTransformerTest {
       new JsonObject().put("string", "foo").put("number", 13.37).put("integer", 42).put("boolean", true);
 
     return Stream.of(
-      Arguments.of("empty", DUMMY_PARAM, "", EMPTY_JSON_OBJECT),
-      Arguments.of("empty (exploded)", DUMMY_PARAM_EXPLODE, "", EMPTY_JSON_OBJECT),
-      Arguments.of(complexRaw, DUMMY_PARAM, complexRaw, expectedComplex),
-      Arguments.of(complexExplodedRaw + " (exploded)", DUMMY_PARAM_EXPLODE, complexExplodedRaw,
+      Arguments.of("empty", OBJECT_PARAM, "", EMPTY_JSON_OBJECT),
+      Arguments.of("empty (exploded)", OBJECT_PARAM_EXPLODE, "", EMPTY_JSON_OBJECT),
+      Arguments.of(complexRaw, OBJECT_PARAM, complexRaw, expectedComplex),
+      Arguments.of(complexExplodedRaw + " (exploded)", OBJECT_PARAM_EXPLODE, complexExplodedRaw,
         expectedComplex)
     );
   }
@@ -106,11 +110,11 @@ class SimpleTransformerTest {
 
   @Test
   void testInvalidValues() {
-    assertThrows(DecodeException.class, () -> TRANSFORMER.transformPrimitive(DUMMY_PARAM, "\""));
+    assertThrows(DecodeException.class, () -> TRANSFORMER.transformPrimitive(STRING_PARAM, "\""));
 
     String invalidObject = "string,foo,number";
     ValidatorException exception =
-      assertThrows(ValidatorException.class, () -> TRANSFORMER.transformObject(DUMMY_PARAM, invalidObject));
+      assertThrows(ValidatorException.class, () -> TRANSFORMER.transformObject(OBJECT_PARAM, invalidObject));
     String expectedMsg = "The formatting of the value of path parameter dummy doesn't match to style simple.";
     assertThat(exception).hasMessageThat().isEqualTo(expectedMsg);
   }

